### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by Guillem-db**
First round of feedback on the technical report.

### CODE REVIEW

File: technical_report.md [hunk 1]
```diff
@@ -2,20 +2,130 @@
...
+
+
+# Introduction
+
+Fractal Monte Carlo (FMC) is a new model-free reinforcement learning (MFRL)[#Citation Needed]()
+agent derived from first principles of the Fractal AI theory[#CN](). FMC calculates an approximation
+to a probability distribution over the action space that acts as an utility function. In this
+report, we will show how our method outperforms equivalent approaches based on discounting expected
+rewards[#CN]() to construct a policy.
+
```
Comment (eiso) [inline technical_report.md]: Fractal Monte Carlo (FMC) is a new model-free RL agent derived from first principles based upon the FAI theory. FMC contains a utility function that approximates the probability distribution over the potential action space. And builds a policy by discounting expected rewards. This report explains how.

File: technical_report.md [hunk 2]
```diff
@@ -2,20 +2,130 @@
...
+agent derived from first principles of the Fractal AI theory[#CN](). FMC calculates an approximation
+to a probability distribution over the action space that acts as an utility function. In this
+report, we will show how our method outperforms equivalent approaches based on discounting expected
+rewards[#CN]() to construct a policy.
+   
+FMC has been extensively tested in a wide range of environments, in both continuous and discrete 
+action spaces. This report will focus on the performance
+of the Agent in Atari games, as provided by the Python library OpenAI Gym[#CN](), under heavily
+bounded computational resources. The agent presented outperforms the current state of the art
+and shows to be significantly more efficient than similar methods, using either pixels or RAM as input.
```
Comment (eiso) [inline technical_report.md]: s/outperforms the current state of the art/outperforms the current state of the art on X/Y Atari games./

File: technical_report.md [hunk 3]
```diff
@@ -2,20 +2,130 @@
...
+
+# How it works
+
+FMC uses a [Swarm](#swarm) to construct a [causal cone](#causal-cone) of possible future
```
Comment (eiso) [inline technical_report.md]: Can we explain here it is a tree (and call it a causal cone in brackets), especially when using the term leaf nodes next?

Comment (eiso) [reply technical_report.md]: e.g. In FMC the tree of potential actions on the states from t0 to tn is called a causal cone.

File: technical_report.md [hunk 4]
```diff
@@ -2,20 +2,130 @@
...
+trajectories of the agent through the state space. The leaf nodes in this cone, will be used
+to estimate the the utility value distribution over actions. This utility distribution will be used
```
Comment (eiso) [inline technical_report.md]: s/the the/the
s/in this clone/in this tree (causal cone)

You need to explain what values in the lead nodes will be used to calculate the utility value distribution

File: technical_report.md [hunk 5]
```diff
@@ -2,20 +2,130 @@
...
+as a policy, and the highest utility action will be chosen.
+
+The causal cone is calculated using an iterative algorithm that makes random perturbations
+to a group of swarm. FMC intends to distribute the swarm of the Swarm matching de reward density
+distribution of the state space sampled.
+
```
Comment (eiso) [inline technical_report.md]: The causal cone, unlike MCTS is not a static tree of all possible actions. Instead the causal cone is a tree data structure that changes at every time step by applying random perturbations and evaluation the reward density distribution.

File: technical_report.md [hunk 6]
```diff
@@ -2,20 +2,130 @@
...
+This is accomplished thanks to the following properties of the algorithm:
+
+- FMC uses an scalar called [virtual reward](#virtual-reward) to weight both the physical
```
Comment (eiso) [inline technical_report.md]: virtual reward is X * Y... don't make people go to the definition section unnecessarily

File: technical_report.md [hunk 7]
```diff
@@ -2,20 +2,130 @@
...
+distribution of the Swarm across the state space, and the reward of the different swarm.
+
+- The algorithm recycles useless trajectories in the cone by [cloning](#cloning) any state of the
```
Comment (eiso) [inline technical_report.md]: trajectories should always be called paths. That's the computer science accepted terminology when talking about tree data structures.

File: technical_report.md [hunk 8]
```diff
@@ -2,20 +2,130 @@
...
+Swarm to another randomly chosen state. An state will clone or not depending on the relationship 
+of virtual rewards between itself and the chosen state.
+
```
Comment (eiso) [inline technical_report.md]: overly complex explanation. Think about how you would explain it to a novice. If you need help here, let me know.

File: technical_report.md [hunk 9]
```diff
@@ -2,20 +2,130 @@
...
+- After calculating an action, the algorithm updates most of its internal parameters to adjust
+itself dynamically to changes in the environment. The corrections to the parameters needed,
+are calculated based on how asymmetric the shape of the swarm is. The asymmetry is measured with
+respect to the spatial distribution of the swarm, their reward distribution, and their distribution
+across time.
```
Comment (eiso) [inline technical_report.md]: After calculating an action where... e.g. at each time step (next state) an action is determined by ... once an action is determined, the potential paths to explore the tree are updated ... To update them....

File: technical_report.md [hunk 10]
```diff
@@ -2,20 +2,130 @@
...
+
```
Comment (eiso) [inline technical_report.md]: The below is a different section. Above explains how, while this explains benefits:

File: technical_report.md [hunk 11]
```diff
@@ -2,20 +2,130 @@
...
+- It can discard arbitrary large sections of the search space by manually adding boundary conditions.
+These boundary conditions are modeled defining a [dead flag]() for each state, that allows to modify the
+cloning probability.
+
+- FMC avoids getting stuck in local optima thanks to weighting not only the reward, but also the
+spatial distribution of the states in the swarm.
```
Comment (eiso) [inline technical_report.md]: Try to phrase this without the foreign concept of a swarm

File: technical_report.md [hunk 12]
```diff
@@ -2,20 +2,130 @@
...
+
+- After calculating an action, the algorithm updates most of its internal parameters to adjust
+itself dynamically to changes in the environment. The corrections to the parameters needed,
+are calculated based on how asymmetric the shape of the swarm is. The asymmetry is measured with
+respect to the spatial distribution of the swarm, their reward distribution, and their distribution
+across time.
+
+- It can discard arbitrary large sections of the search space by manually adding boundary conditions.
+These boundary conditions are modeled defining a [dead flag]() for each state, that allows to modify the
+cloning probability.
```
Comment (eiso) [inline technical_report.md]: > These boundary conditions are modeled defining a [dead flag]() for each state, that allows to modify the cloning probability.

This should be in the how, not the benefits. And should be better explained.

File: technical_report.md [hunk 13]
```diff
@@ -2,20 +2,130 @@
...
+
+- FMC avoids getting stuck in local optima thanks to weighting not only the reward, but also the
+spatial distribution of the states in the swarm.
+
+- The algorithm is parallelizable and could scale almost linearly with the maximum number of
```
Comment (eiso) [inline technical_report.md]: almost linear is not a thing, super linear?

File: technical_report.md [hunk 14]
```diff
@@ -2,20 +2,130 @@
...
+samples allowed. All the internal operations (distances, virtual reward calculations, cloning,
+etc...) involve only one state, or a given state and another randomly chosen state. 
+They all have a computational complexity that is lower than quadratic, although the specific
+scalability will depend on the implementation.
```
Comment (eiso) [inline technical_report.md]: quadratic too what? # of paths being evaluated? walkers?

File: technical_report.md [hunk 15]
```diff
@@ -2,20 +2,130 @@
...
+
+- It is tolerant to rewards and distances that are unbounded, or vary widely in scale thanks to a
```
Comment (eiso) [inline technical_report.md]: normalization should be explained above in the how. This is a benefit section..

File: technical_report.md [hunk 16]
```diff
@@ -2,20 +2,130 @@
...
+
+##### Causal cone
+ It represents the graph formed by all the paths traversed by the swarm of the
+ swarm during the evolutionary process that they undergo. All the paths share a common origin and
+ form a graph with a tree like structure. At any moment, the swarm of the swarm will be distributed
+ across the leaf nodes of the tree.
+
+##### Swarm
+It is the mathematical structure that represents a group of swarm that undergo an
+evolution process through random perturbations. [See lines 58, 105]()
```
Comment (eiso) [inline technical_report.md]: Either explain this better above, not just in the definition or find a better name/analogy.

File: technical_report.md [hunk 17]
```diff
@@ -2,20 +2,130 @@
...
+ 
+ This mechanism allows for efficient scanning when facing boundary conditions.
+ [See lines 208, 220, 221]()
+    
+##### Distance
+ This is an informative measure of how different the observations of two different
+swarm are. We call it distance because in this specific implementation we are using the euclidean
+distance between two observations. At each iteration of the algorithm, each state will calculate its distance
+to another state chosen at random. [Defined at L162 fractalai.py]()
+
```
Comment (eiso) [inline technical_report.md]: Distance measure should be explained in the how to section as well, not just linked. Your document is full of places where people need to jump from the top to the bottom and back. That is a terrible experience.

File: technical_report.md [hunk 18]
```diff
@@ -2,20 +2,130 @@
...
+
+##### Normalization
+It is a transformation applied to both the distances and the rewards, which is applied before
+calculating the virtual reward. It consists in two steps:
+
+  - Normalizing the distances of all the states so they fall in the range [0,1]. The lowest distance
+will be assigned a value of 0 and the highest a value of one. [L173]()
```
Comment (eiso) [inline technical_report.md]: Doesn't need a definition, you can just say above, we normalize the distance between [0,1]

**Comment by Guillem-db** [comment]
Remember to fix the docstrings.

**Pull request merged**